activiti-cloud-query-service/activiti-cloud-services-query/activiti-cloud-services-query-rest/src/main/java/org/activiti/cloud/services/query/events/handlers/ProcessCompletedEventHandler.java

```diff
@@ -15,44 +15,62 @@
-        this.entityManager = entityManager;
-    }
+  public ProcessCompletedEventHandler(EntityManager entityManager) {
+    this.entityManager = entityManager;
+  }
 
-    @Override
-    public void handle(CloudRuntimeEvent<?, ?> event) {
-        CloudProcessCompletedEvent completedEvent = (CloudProcessCompletedEvent) event;
-        String processInstanceId = completedEvent.getEntity().getId();
-        Optional<ProcessInstanceEntity> findResult = Optional.ofNullable(entityManager.find(ProcessInstanceEntity.class,
-                                                                                            processInstanceId));
-        if (findResult.isPresent()) {
-            ProcessInstanceEntity processInstanceEntity = findResult.get();
-            processInstanceEntity.setStatus(ProcessInstance.ProcessInstanceStatus.COMPLETED);
-            processInstanceEntity.setLastModified(new Date(completedEvent.getTimestamp()));
-            processInstanceEntity.setCompletedDate(new Date(completedEvent.getTimestamp()));
-            entityManager.persist(processInstanceEntity);
-        } else {
-            throw new QueryException("Unable to find process instance with the given id: " + processInstanceId);
-        }
-    }
+  @Override
+  public void handle(CloudRuntimeEvent<?, ?> event) {
+    CloudProcessCompletedEvent completedEvent = (CloudProcessCompletedEvent) event;
+    String processInstanceId = completedEvent.getEntity().getId();
+    Optional<ProcessInstanceEntity> findResult = Optional.ofNullable(entityManager.find(ProcessInstanceEntity.class,
+                                                                                        processInstanceId));
+    if (findResult.isPresent()) {
+      ProcessInstanceEntity processInstanceEntity = findResult.get();
+      processInstanceEntity.setStatus(ProcessInstance.ProcessInstanceStatus.COMPLETED);
+      processInstanceEntity.setLastModified(new Date(completedEvent.getTimestamp()));
+      processInstanceEntity.setCompletedDate(new Date(completedEvent.getTimestamp()));
+      markAllChildTasksAsCancelledWhenTheyAreAssignedOrCreated(processInstanceEntity);
+      entityManager.persist(processInstanceEntity);
 
-    @Override
-    public String getHandledEvent() {
-        return ProcessRuntimeEvent.ProcessEvents.PROCESS_COMPLETED.name();
+    } else {
+      throw new QueryException("Unable to find process instance with the given id: " + processInstanceId);
     }
+  }
+
+  private void markAllChildTasksAsCancelledWhenTheyAreAssignedOrCreated(ProcessInstanceEntity processInstanceEntity) {
+    Predicate<TaskEntity> cancellableTasks = task -> TaskStatus.ASSIGNED.equals(task.getStatus())
+        || TaskStatus.CREATED.equals(task.getStatus());
+
+    Stream.ofNullable(processInstanceEntity.getTasks())
+        .flatMap(Set::stream)
+        .filter(cancellableTasks)
+        .peek(task -> task.setStatus(TaskStatus.CANCELLED))
```

Comment:
I'm not sure if this solution will be stable in long period. Task cancelling having dedicated event handler so maybe we could create a new task canceled event with those tasks?
It might help for situation when task cancel logic would change.

Reply:
This sounds good but, do you think you are able to call the handler for cancelled tasks from here?

Reply:
See two options:
1. Use TaskCancelledEventHandler, just create event and let it handle it (dependency have to be set in spring config) 
2. Just publish message to 'queryConsumer' which will be routed and then handled by specific handler class

IMHO better option might be 1. as we don't have many readers for the same event

Reply:
I don't like creating an event for this, because it is a "fake" event, is not an event produced by the RB and can be confusing if we have a bridge or some other service listening to the topic where we will be placing the event.

Is it possible to call the TaskCancelledEventHandler directly? (I think this is what you meant by option 1 right?) If it is, then I would go with this approach.

Reply:
ok, change applied, cannot double check at this moment (https://hxpmock-apa.envalfresco.com/ is down for deployments)

Reply:
@pmartinezga @graric Tests were successful